Distributed cell (distribcell) tally filters

wbinventor opened the pull request:
This PR introduces distributed cell (distribcell) tally filters into OpenMC. Presently, when a "cell" filter is specified for a tally, particles traveling through any instance of that cell will be tallied into the same location (*e.g.*, a fuel pin cell placed many times throughout a geometry). In contrast, "distribcell" filters enable users to compute tallies for each unique **instance** of a cell throughout the geometry.

The algorithm for distribcell tallies is a generalization of the algorithm developed and applied to identify flat source regions in constructive solid geometries (CSG) for OpenMOC. The basic premise of the algorithm is to build **offset tables** at each level of a CSG tree as a pre-processing step. The **instance ID** for each cell instance can be rapidly computed on the fly for a given nested particle coordinates list by traversing the tree and summing up the appropriate offsets from the offset tables at each level. A detailed explanation of the algorithm, including pseudocode as well as a performance model for the memory footprint can be found in the following paper:

* D. Lax, W. Boyd and N. Horelik. "An Algorithm for Identifying Unique Regions in Constructive Solid Geometries." *Proc. of PHYSOR*, Kyoto, Japan (2014).

A user can simply place a distribcell" filter on a tally in the same way as a typical "cell" filter:

```xml
...

    <tally id="1">
        <filter bins="101" type="distribcell" />
        <scores>flux</scores>
    </tally>

...
```

At the beginning of a simulation, OpenMC will count the number of unique instances of the cell, allocate an array for the tally bins for each instance. During the simulation, each event occurring within a distributed cell will be incremented into the appropriate tally bin(s) based on the instance ID for the location of the particle.

The majority of the development work for distribcell tallies was done by @laxderek. Both @nhorelik and I have used this functionality on the closed source repo for over a year now and believe that it is ready for use by the broader community now. This functionality forms the basis for unstructured mesh tallies needed for large scale tallies, and is a prerequisite for @nhorelik's coming spatial domain decomposition PR.

wbinventor commented:
I should clarify that distribcell tallies have been implemented on the closed source side for quite some time. Since @laxderek graduated, I went through and updated this functionality to work with hexagonal lattices, and to export the appropriate metadata (offset tables, maps) to the HDF5 summary files for tally post-processing.

smharper commented:
There's a slight wrinkle that's introduced by #359.  Any universe that appears in a lattice `outer` element could be replicated a bunch of times, but we don't easily know how many times.  In fact, such a universe can very easily be replicated an infinite number of times (or 2^32^3 if you want to be pedantic).

I would recommend adding some code to input_xml.F90 that would forbid a user from applying a distribcell filter to any cell in any universe that appears in a lattice's `outer` element.

wbinventor commented:
@smharper - this is a very good point. In reality, it would be very nice to count cell instances that are in the ``outer`` universes of a lattice, but I do not see a good general way to go about this with the current implementation of ``outer``. My attempt to get around this was to ignore instances in ``outer`` universes specified on lattices. In particular, during particle tracking the distribcell mapping is only computed when your new ``is_valid_indices(...)`` routine returns true ([link](https://github.com/mit-crpg/openmc/pull/379/files#diff-6b51ef6c329326abb8f8ee71c9d8cec3R266)). As I understand, this routine will return false for locations in the ``outer`` universe of a lattice. I recognize that this may not be transparent to the user, so if you and the rest of the team thinks I should put some error checking code in ``input_xml.F90`` to do as you indicated, I'll go ahead with it.

wbinventor commented:
Also for the record, I can't for the life of me figure out why the tests are failing on Travis. I've tried to reproduce the errors on my system to no avail. Is there anyway to download the logs from Travis so I can see what is going awry?

paulromano commented:
The travis VMs use Ubuntu 12.04 which in turn uses gcc 4.6. When I try to compile with gfortran 4.6 I get the following error message:
```
/home/romano/openmc/src/eigenvalue.F90:77.50:

!$omp parallel do schedule(static) firstprivate(p)
                                                  1
Error: FIRSTPRIVATE clause object 'p' has ALLOCATABLE components at (1)
```

This may have been a restriction in a previous version of the OpenMP standard which no longer applies (hence why later versions of gfortran accept it with no error or warning). As long as Travis is using 12.04, we are going to stick to gfortran 4.6 compatibility though, so this will need to be fixed.

wbinventor commented:
Thanks for looking into the Travis build @paulromano. This is indeed a tricky one, esp considering that I'm a Fortran novice. The issue appears to be the result of the new [``mapping``](https://github.com/mit-crpg/openmc/pull/379/files#diff-54c34c0d33e0989aa2d7cfdb0138af7fR31) allocatable integer array for the ``LocalCoord`` derived type. If anyone has any thoughts on how to define this in an alternative way so that the old OpenMP standard respects it, I'd certainly appreciate it.

wbinventor commented:
The OMP issue with the gfortran version 4.6 on Travis has been solved - all tests now pass.

smharper commented on src/geometry.F90:
```diff
@@ -129,8 +129,8 @@ recursive subroutine find_cell(p, found, search_cells)
     type(Particle), intent(inout) :: p
     logical,        intent(inout) :: found
     integer,        optional      :: search_cells(:)
-
```
I would prefer that you leave the blank space here to separate dummy variables from regular variables.

smharper commented on src/geometry.F90:
```diff
@@ -568,8 +588,8 @@ subroutine cross_lattice(p, lattice_translation)
 
     type(Particle), intent(inout) :: p
     integer,        intent(in)    :: lattice_translation(3)
-
```
same comment

smharper commented on src/geometry_header.F90:
```diff
@@ -7,13 +7,15 @@ module geometry_header
+     real(8) :: y0                     ! Translation in y-coordinate
+     real(8) :: z0                     ! Translation in z-coordinate
+     integer, allocatable :: kount(:)  ! Distribcell target counts
+     logical, allocatable :: search(:) ! Distribcell search completed
```
Can we make the `kount` and `search` arrays local variables in the `prepare_distribcell` subroutine?  I don't like carrying them around throughout the entire simulation when we only need them during initialization (even though they are deallocated).  I think we could pass the `kount` and `search` arrays as arguments to `calc_offsets`.

smharper commented on src/output.F90:
```diff
@@ -2162,5 +2184,244 @@ function get_label(t, i_filter) result(label)
+        ! AT LOWEST UNIVERSE, TERMINATE SEARCH
+        !if (c % type == CELL_NORMAL) then
+        !  call fatal_error("Unexpected end of search with normal cell.")
+
```
Remove the commented out code

wbinventor commented:
@smharper - thanks for the review. I just updated this PR to reflect each of your comments, most notably removing the ``kount`` and ``search`` variables from the ``Universe`` derived type and making them local to ``prepare_distribcell()`` routine in "initialize.F90". In addition, I renamed them to ``kounts`` and ``found``, respectively.

smharper commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+
+      if (extra > 0) then
+        call fatal_error("At least one Distribcell filter was specified with&
+                         &multiple bins. This feature is not yet supported.")
```
Need an extra space between "with" and "multiple"

smharper commented on src/output.F90:
```diff
@@ -2162,5 +2184,239 @@ function get_label(t, i_filter) result(label)
+            ! Loop over lattice coordinates
+            do k = 1, n_x
+              do l = 1, n_y
+                do m = 1, n_z
```
Reverse the order of these do loops.  I know this code isn't really performance critical, but it's probably still best to consistently use the fast ordering.

wbinventor replied:
Yes I agree

smharper commented on src/output.F90:
```diff
@@ -2162,5 +2184,239 @@ function get_label(t, i_filter) result(label)
+
+            n_x = lat % n_axial
+            n_y = 2 * lat % n_rings - 1
+            n_z = 2 * lat % n_rings - 1
```
`n_x = 2 * lat % n_rings - 1`
`n_z = lat % n_axial`

wbinventor replied:
Oops

wbinventor commented:
I just pushed the changes that you requested @smharper

smharper commented on src/output.F90:
```diff
@@ -2162,5 +2184,239 @@ function get_label(t, i_filter) result(label)
+                    next_univ => universes(lat % universes(old_k, old_l, old_m))
+                    path = trim(path) // "(" // trim(to_str(old_k)) // &
+                         "," // trim(to_str(old_l)) // "," // &
+                         trim(to_str(old_m)) // ")"
```
It's not very transparent to the user what these indices mean.  I think we should output `old_k - lat % n_rings` and `old_l - lat % n_rings`.  This will make the center position (0, 0).  Then we can add a bit to the distribcell note in the docs that references the methods section 2.4 which shows how our indexing works.

smharper replied:
i.e. http://web.mit.edu/smharper/www/methods/geometry.html#finding-a-lattice-tile

wbinventor replied:
Yes, this seems reasonable enough to me. In reality, I'll go with whatever indexing scheme you think is best since you're the mastermind behind hex lattices. I simply was trying to figure out a way to print a semi-useful "path" for each distribcell to "tallies.out" here, so this is just the simplest scheme I knew how to do for hex lattices. I just revised as you suggested and pushed. You might try running "examples/python/lattice/hexagonal/build-xml.py" to see if the paths that are printed out are as you expect / wish.

Also, I really, really like your writeup on lattice indexing in OpenMC. Do you plan to submit that as a PR for the main website?

smharper replied:
That documentation that I linked is already pulled into the develop branch.  It just doesn't show on the GitHub documentation because that is the documentation for the latest release, not the develop branch.

smharper commented on src/output.F90:
```diff
@@ -2162,5 +2184,239 @@ function get_label(t, i_filter) result(label)
+                    next_univ => universes(lat % universes(k, l, m))  
+                    path = trim(path) // "(" // trim(to_str(k)) // &
+                         "," // trim(to_str(l)) // "," // &
+                         trim(to_str(m)) // ")"
```
The `k`, `l`, and `m` above should be `old_k`, `old_l`, and `old_m`, right?

wbinventor replied:
Actually, this whole block for ``RectLattice`` needed some fixing, which I believe I accomplished with my latest commit. To be perfectly honest, I never use "tallies.out" and only process tallies with the statepoint files. So I haven't fully vetted the output of distribcell path strings for complicated geometries. I've run some cases with both simple and nested lattices (both rectangular and hexagonal) and things are everything now appears as I expect it should.

smharper commented on src/output.F90:
```diff
@@ -2162,5 +2184,239 @@ function get_label(t, i_filter) result(label)
+        end if
+      end if
+
+      if (n == 1 .and. c % type /= 1) then
```
`c % type /= CELL_NORMAL`

wbinventor replied:
Yes, good catch.

smharper commented:
The logic of `find_offset` is pretty convoluted.  I'd like to take a crack at simplifying it before we merge this, but I should probably be studying for the 22.211 final on Monday.  Poke me if you don't hear anything else from me by Tuesday evening.

wbinventor commented:
Yes, I agree that ``find_offset`` is pretty miserable. It was the bane of my existence in patching this PR together from @laxderek's original work - you'd be surprised that it is actually much better than before. I'm unsure how to clean it up, but if you have some ideas, I'll be happy to help.

wbinventor commented:
And I once again have no idea why the test suite is failing on Travis now.  The log doesn't indicate what the error is - it seems to be an issue during the build and the output is suppressed. Everything works fine on my system with gfortran versions 4.6 and 4.8, so I'm not sure what's wrong.

Related to this, perhaps we should look into a way to output the full log from Travis so that we have a clearer idea of what is going on when things break during the build rather than at runtime.

paulromano commented:
The `travis_install.sh` script was failing because a new release of HDF5 (1.8.15) caused the URL where HDF5 was being downloaded from to become broken. I fixed the URL on the develop branch, so hopefully this should fix the problem. I just restarted the Travis build.

wbinventor commented:
Awesome, that seems to have fixed the issue - thanks @paulromano!

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+        ! Determine type of filter
+        if (tally % filters(j) % type == FILTER_DISTRIBCELL) then
+          count_all = .true.
+          extra = extra + size(tally % filters(j) % int_bins) - 1
```
Why not just do an explicit check for `size(tally % filters(j) % int_bins) > 0` here and raise a fatal error if it's true?

wbinventor replied:
Oh wow, very good point, I have no idea why the error check was written in this way. Just removed the ``extra`` variable and check for this situation in the way that you suggest.

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+    type(Universe),       pointer :: univ             ! Pointer to universe
+    type(Cell),           pointer :: c                ! Pointer to cell
+    integer, allocatable :: univ_list(:)              ! Target offsets
+    integer, allocatable :: kounts(:,:)               ! Target count
```
`counts` is not an intrinsic procedure or reserved word in Fortran, so I'd much prefer it over `kounts`.

wbinventor replied:
Absolutely, I kept this variable name around but should have changed it before opening the PR. Both ``kount`` and ``kounts`` are ``count`` and ``counts`` in "initialize.F90" and "geometry.F90" now.

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+  end subroutine prepare_distribcell
+
+
+  
```
One empty line separating a subroutine and the next comment block is the prevailing style in OpenMC code.

wbinventor replied:
Got it, thanks.

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+        if (cell_list % has_key(univ % cells(j))) then
+          
+            ! Loop over all tallies    
+            do l = 1, n_tallies
```
Extra/inconsistent indentation here

wbinventor replied:
Fixed

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+          
+          univ_list(k) = univ % id
+          k = k + 1
+          cycle
```
`cycle` doesn't do anything here since it's at the end of the loop (?)

wbinventor replied:
Good call, this is a residual from the implementation before I restructured - I just removed it.

paulromano commented on src/geometry.F90:
```diff
@@ -1582,5 +1614,327 @@ subroutine handle_lost_particle(p, message)
+      end if
+    end do
+    
+    return
```
No need for `return`

wbinventor replied:
Ditto

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +916,230 @@ subroutine allocate_banks()
+    logical, intent(out), allocatable     :: found(:,:)   ! Target found
+
+    integer :: i, j, k, l, m                    ! Loop counters
+    type(DictIntInt) :: cell_list               ! distribells to track    
```
It doesn't look like the values of the dictionary are used (set to zero for all keys). You should use `type(SetInt)` from set_header.F90 instead.

wbinventor replied:
Sure, using ``SetInt`` now.

paulromano commented:
Need to update RELAX NG schema for tallies.xml in `src/relaxng/tallies.rnc`

wbinventor commented:
I just added the "distribcell" filter type to the RELAX NG schema. @paulromano - all of your comment have been addressed.

paulromano commented on src/geometry.F90:
```diff
@@ -1582,5 +1614,327 @@ subroutine handle_lost_particle(p, message)
+              do m = 1, lat % n_cells(3)
+                lat % offset(map, j, k, m) = offset
+                next_univ => universes(lat % universes(j, k, m))
+                offset = offset+count_target(next_univ,kounts,found,goal,map)
```
Procedure arguments should be separated by a [single space](http://mit-crpg.github.io/openmc/devguide/styleguide.html#whitespace-in-expressions)

paulromano commented on src/geometry.F90:
```diff
@@ -1582,5 +1614,327 @@ subroutine handle_lost_particle(p, message)
+                else
+                  lat % offset(map, j, k, m) = offset
+                  next_univ => universes(lat % universes(j, k, m))
+                  offset = offset+count_target(next_univ,kounts,found,goal,map)
```
same comment re: spacing

paulromano commented on src/geometry.F90:
```diff
@@ -198,6 +199,14 @@ recursive subroutine find_cell(p, found, search_cells)
+        end if
+        do j = 1, n_maps
+          p % coord % mapping(j) = c % offset(j)
+        end do
```
This could just be `p % coord % mapping(:) = c %  offset(:)`

wbinventor replied:
Oh wow, this what happens when a C++ programmer tries Fortran

paulromano commented on src/geometry.F90:
```diff
@@ -253,6 +262,18 @@ recursive subroutine find_cell(p, found, search_cells)
+
+          do j = 1, n_maps
+            p % coord % mapping(j) = lat % offset(j, i_xyz(1), i_xyz(2), i_xyz(3))
+          end do
```
Maybe use `p % coord % mapping(:) = lat % offset(:, i_xyz(1), i_xyz(2), i_xyz(3))`

paulromano commented on src/geometry.F90:
```diff
@@ -612,8 +634,18 @@ subroutine cross_lattice(p, lattice_translation)
+      end if
+      do i = 1, n_maps
+        p % coord % mapping(i) = lat % offset(i, i_xyz(1), i_xyz(2), i_xyz(3))
+      end do
```
same here

paulromano commented on src/geometry.F90:
```diff
@@ -1582,5 +1614,327 @@ subroutine handle_lost_particle(p, message)
+    logical, intent(inout)        :: found(:,:)   ! Target found
+    integer, intent(in) :: goal            ! target universe ID
+    integer             :: kount           ! number of times target located
+    integer             :: map             ! current map
```
Specify intent on `map`

wbinventor replied:
Good point, just fixed this and cleaned up the variable declarations for ``calc_offsets`` and ``count_target``

paulromano commented on src/geometry.F90:
```diff
@@ -1582,5 +1614,325 @@ subroutine handle_lost_particle(p, message)
+    logical, intent(inout)        :: found(:,:)   ! Target found
+    integer, intent(in) :: goal            ! target universe ID
+    integer             :: count           ! number of times target located
+    integer             :: map             ! current map
```
Specify intent on `map`

paulromano commented on src/geometry_header.F90:
```diff
@@ -28,6 +28,8 @@ module geometry_header
     integer              :: outside          ! Material to fill area outside
     integer              :: outer            ! universe to tile outside the lat
     logical              :: is_3d            ! Lattice has cells on z axis
+    integer :: level                         ! Level of lattice
```
I'm not seeing where this is used. Can it be deleted?

wbinventor replied:
Nope, yet another residual from a previous era

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +917,218 @@ subroutine allocate_banks()
+      do j = 1, univ % n_cells
+        if (cell_list % contains(univ % cells(j))) then
+          n_maps = n_maps + 1
+          cycle
```
Another instance of `cycle` not doing anything

wbinventor replied:
wow, now that was just plain embarrassing

paulromano commented on src/initialize.F90:
```diff
@@ -902,4 +917,218 @@ subroutine allocate_banks()
+        counts(i,j) = 0
+        found(i,j) = .false.
+      end do
+    end do
```
```Fortran
counts(:,:) = 0
found(:,:) = .false.
```

paulromano commented:
@smharper Were you still planning on reworking the `find_offset` subroutine?

wbinventor commented:
This is up-to-date wrt your comments @paulromano

paulromano commented:
Great. I'm pretty happy with the code as is. Implementation looks clean (although it took me a while to wrap my brain around all the offsets and indexing), and memory requirements are low. No undue impact on performance from what I've observed. I'm ok with merging now, pending the resolution of potential changes to `find_offset`. @smharper if you want to take a stab now, that's fine. If you want to wait until later and submit another PR, that's also fine.

paulromano commented on tests/test_filter_distribcell/test_filter_distribcell.py:
```diff
@@ -0,0 +1,112 @@
+    output.append(cwd + '/geometry.xml')
+    output.append(cwd + '/materials.xml')
+    output.append(cwd + '/settings.xml')
+    output.append(cwd + '/tallies.xml')
```
What's the purpose of deleting xml files during teardown?

wbinventor replied:
That is a very good question - I'm not sure why @laxderek implemented the
test that way. I'll see if I can go ahead and remove this later today.
On May 27, 2015 5:28 AM, "Paul Romano" <notifications@github.com> wrote:

> In tests/test_filter_distribcell/test_filter_distribcell.py
> <https://github.com/mit-crpg/openmc/pull/379#discussion_r31127759>:
>
> > +    statepoint.append(glob.glob(cwd + '/case-1/statepoint.1.*'))
> > +    statepoint.append(glob.glob(cwd + '/case-2/statepoint.1.*'))
> > +    statepoint.append(glob.glob(cwd + '/case-3/statepoint.3.*'))
> > +    call(['python', 'results.py', statepoint.pop()[0], statepoint.pop()[0], statepoint.pop()[0]])
> > +    compare = filecmp.cmp('results_test.dat', 'results_true.dat')
> > +    if not compare:
> > +      os.rename('results_test.dat', 'results_error.dat')
> > +    assert compare
> > +
> > +def teardown():
> > +    cwd = os.getcwd()
> > +    output = glob.glob(cwd + '/statepoint.*')
> > +    output.append(cwd + '/geometry.xml')
> > +    output.append(cwd + '/materials.xml')
> > +    output.append(cwd + '/settings.xml')
> > +    output.append(cwd + '/tallies.xml')
>
> What's the purpose of deleting xml files during teardown?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/379/files#r31127759>.
>

laxderek replied:
I'm not sure that the file deletion was necessary, I honestly don't recall
at this point. I'm tempted to say it had something to do with the first
test not being run correctly due to the old xml files being present, but if
that is not the case, then there is no need for deletion.

On Wed, May 27, 2015 at 1:09 PM, Will Boyd <notifications@github.com> wrote:

> In tests/test_filter_distribcell/test_filter_distribcell.py
> <https://github.com/mit-crpg/openmc/pull/379#discussion_r31156620>:
>
> > +    statepoint.append(glob.glob(cwd + '/case-1/statepoint.1.*'))
> > +    statepoint.append(glob.glob(cwd + '/case-2/statepoint.1.*'))
> > +    statepoint.append(glob.glob(cwd + '/case-3/statepoint.3.*'))
> > +    call(['python', 'results.py', statepoint.pop()[0], statepoint.pop()[0], statepoint.pop()[0]])
> > +    compare = filecmp.cmp('results_test.dat', 'results_true.dat')
> > +    if not compare:
> > +      os.rename('results_test.dat', 'results_error.dat')
> > +    assert compare
> > +
> > +def teardown():
> > +    cwd = os.getcwd()
> > +    output = glob.glob(cwd + '/statepoint.*')
> > +    output.append(cwd + '/geometry.xml')
> > +    output.append(cwd + '/materials.xml')
> > +    output.append(cwd + '/settings.xml')
> > +    output.append(cwd + '/tallies.xml')
>
> That is a very good question - I'm not sure why @laxderek
> <https://github.com/laxderek> implemented the test that way. I'll see if
> I can go ahead and remove this later today.
> … <#14d96596817abce9_>
> On May 27, 2015 5:28 AM, "Paul Romano" <notifications@github.com> wrote:
> In tests/test_filter_distribcell/test_filter_distribcell.py <
> https://github.com/mit-crpg/openmc/pull/379#discussion_r31127759>: > +
> statepoint.append(glob.glob(cwd + '/case-1/statepoint.1.*')) > +
> statepoint.append(glob.glob(cwd + '/case-2/statepoint.1.*')) > +
> statepoint.append(glob.glob(cwd + '/case-3/statepoint.3.*')) > +
> call(['python', 'results.py', statepoint.pop()[0], statepoint.pop()[0],
> statepoint.pop()[0]]) > + compare = filecmp.cmp('results_test.dat',
> 'results_true.dat') > + if not compare: > + os.rename('results_test.dat',
> 'results_error.dat') > + assert compare > + > +def teardown(): > + cwd =
> os.getcwd() > + output = glob.glob(cwd + '/statepoint.*') > +
> output.append(cwd + '/geometry.xml') > + output.append(cwd +
> '/materials.xml') > + output.append(cwd + '/settings.xml') > +
> output.append(cwd + '/tallies.xml') What's the purpose of deleting xml
> files during teardown? — Reply to this email directly or view it on GitHub <
> https://github.com/mit-crpg/openmc/pull/379/files#r31127759>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/379/files#r31156620>.
>

wbinventor replied:
Looks like we (no longer) need this, so I just removed it.

paulromano commented:
I can't quite figure out why all the tests are failing on the phdf5-omp-debug configuration. It works fine locally on my machine. Can you try updating the version of HDF5/PHDF5 in tests/travis_install.sh to 1.8.15 and see if that fixes it?

nelsonag commented:
Is it because the python API currently requires opencg?  (also where do I  get that...)

paulromano commented:
Nope, in this case, it doesn't have anything to do with OpenCG. OpenCG hasn't been openly released yet. @wbinventor can probably comment further.

nelsonag commented:
Right, I should have read on to your comment pegging it to only the
phdf5-omp-debug run.  I was reacting to how summary.py pulls in
openmc.opencg_compatible without checking it available first.

On Thu, May 28, 2015 at 8:50 AM, Paul Romano <notifications@github.com>
wrote:

> Nope, in this case, it doesn't have anything to do with OpenCG. OpenCG
> hasn't been openly released yet. @wbinventor
> <https://github.com/wbinventor> can probably comment further.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/379#issuecomment-106300018>.
>

wbinventor commented:
@paulromano - I just changed the HDF5 version to 1.8.15. I certainly don't understand why any of the minor changes I made with respect to your comments would be an issue, esp. since all tests pass for phdf5-omp-debug with HDF5 v1.8.14 on my machine.

@nelsonag - thanks for pointing out this issue with the OpenCG import. We certainly do need a try-catch around the OpenCG import in "summary.py" and "opencg_compatible.py". I'll add this and open a new PR in just a moment. I don't plan to release OpenCG until the end of the summer, at which point we can discuss whether or not it would useful to add it is a git submodule for OpenMC so that this is never an issue again.

wbinventor commented:
I don't know if the newer version of HDF5 did the trick (or if Travis was just throwing a random fit, which seems to happen sometimes), but the tests now pass.

paulromano commented:
Travis was throwing a random fit -- I just wish I knew the reason and why updating HDF5 caused it to stop having a fit. Anyway, thanks for working through all the issues!
